### PR TITLE
Fix: Duplicate keys in JSON produced for Catalog objects #915

### PR DESCRIPTION
**Pull request by vedansh-5**
## Title
fix: Remove duplicate type field in Catalog serialization

### Problem
The Catalog API responses were showing duplicate `type` fields due to Jackson's polymorphic type handling.

### Solution
- Created custom template for JsonTypeInfo annotation
- Modified serialization to use EXISTING_PROPERTY instead of PROPERTY
- Added comprehensive serialization tests

### Files Changed
1. `api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java`
   - Added tests for Unicode characters
   - Added tests for whitespace handling
   - Added tests for role ARN validation
   - Added tests for null/empty fields

2. `api/management-model/templates/typeInfoAnnotation.mustache` (New)
   - Custom template for JsonTypeInfo annotation
   - Uses EXISTING_PROPERTY strategy

### Testing
- ✅ All unit tests passing
- ✅ Verified with Postman
- ✅ Tested edge cases:
  - Unicode characters
  - Whitespace handling
  - Role ARN validation
  - Null fields
  - Empty fields

### Screenshots
Screenshot 2025-02-06 221503
Screenshot 2025-02-06 221447

### CODE REVIEW

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/README.md
Comment (eric-maynard) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/README.md:1]: In general, I'm against trying to document code like this as it inevitably rots.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    public void testUnicodeCharacters() throws JsonProcessingException {
+        String unicodeName = "测试目录";
+        Catalog catalog = new Catalog(
+                Catalog.TypeEnum.INTERNAL,
+                unicodeName,
+                new CatalogProperties(TEST_LOCATION),
+                null);
+        String json = mapper.writeValueAsString(catalog);
+        Catalog deserialized = mapper.readValue(json, Catalog.class);
+        assertEquals(unicodeName, deserialized.getName());
```
Comment (eric-maynard) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:206]: I like the idea of this kind of round-trip test, but why only check the name?

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        JsonNode jsonNode = mapper.readTree(json);
+
+        assertEquals(1, countTypeFields(jsonNode));
+        assertEquals("INTERNAL", jsonNode.get("type").asText());
+    }
+
+    @Test
+    public void testCatalogDeserialization() throws JsonProcessingException {
+        String json = "{"
+                + "\"type\": \"INTERNAL\","
```
Comment (eric-maynard) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:51]: This format seems much better than the repeated `append` calls below -- and block quotes would probably be even better. Can you try to use this format, or block quotes, below?

File: .vscode/settings.json
Comment (eric-maynard) [inline .vscode/settings.json:1]: This shouldn't be committed; do we need any `.gitignore` changes to help with this? I've not used VS code

**Review by eric-maynard — COMMENTED** [review]
Thanks for the fix! Left a few comments

File: api/management-model/build.gradle.kts [hunk 3]
```diff
@@ -54,12 +57,37 @@ openApiGenerate {
   serverVariables = mapOf("basePath" to "api/v1")
 }
 
+sourceSets {
+    main {
+        java {
+            srcDir(project.layout.buildDirectory.dir("generated/src/main/java"))
```
Comment (dimas-b) [inline api/management-model/build.gradle.kts:63]: Why is this required? I believe gradle should handle that by default.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+        String json = mapper.writeValueAsString(catalog);
+        JsonNode node = mapper.readTree(json);
+        assertEquals(specialName, node.get("name").asText());
+    }
+
+    @Test
+    public void testCatalogWithEmptyStrings() throws JsonProcessingException {
+        String json = "{"
+                + "\"type\": \"INTERNAL\","
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:141]: Do strings with `"""` work?

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/README.md
Comment (vedansh-5) [reply api/management-model/src/test/java/org/apache/polaris/core/admin/model/README.md:1]: @eric-maynard I would love to know better ways to handle it. If you were in my place how would you approach this? Since I'm fairly new to open source any advice about the best practices would be really helpful.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        Catalog catalog = new Catalog(
+                Catalog.TypeEnum.INTERNAL,
+                TEST_CATALOG_NAME,
+                properties,
+                storageConfig);
+
+        String json = mapper.writeValueAsString(catalog);
+        JsonNode jsonNode = mapper.readTree(json);
+
+        assertEquals(1, countTypeFields(jsonNode));
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:44]: I'd prefer to assert exact JSON text. I believe Jackson produces stable output, given the same input, so we can be sure the JSON string will not change from run to run.

**Review by dimas-b — COMMENTED** [review]
Thanks for your contribution, @vedansh-5 !

File: api/management-model/build.gradle.kts [hunk 6]
```diff
@@ -54,12 +57,37 @@ openApiGenerate {
   serverVariables = mapOf("basePath" to "api/v1")
 }
 
+sourceSets {
+    main {
```
Comment (dimas-b) [inline api/management-model/build.gradle.kts:61]: Please run `./gradlew spotlessApply` before pushing to PRs :) CI failed because of that.

File: .vscode/settings.json
Comment (vedansh-5) [reply .vscode/settings.json:1]: Yes, I see my mistake, I will add this into the `.gitignore` file.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/README.md
Comment (eric-maynard) [reply api/management-model/src/test/java/org/apache/polaris/core/admin/model/README.md:1]: I would suggest just adding javadoc comments to your test classes/methods directly so that the code is self-describing. For most tests, a one-line blurb saying what's being tested is more than enough.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java (hunk 4 above)
Comment (vedansh-5) [reply api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:141]: They're only supported in JAVA 15+

File: .vscode/settings.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+    "java.configuration.updateBuildConfiguration": "automatic",
+    "java.compile.nullAnalysis.mode": "automatic",
+    "java.project.sourcePaths": [
```
Comment (gh-yzou) [inline .vscode/settings.json:4]: are we suppose to merge this setting for vscode? if not, we probably should remove this from the pr.

File: server-templates/typeInfoAnnotation.mustache [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    use = JsonTypeInfo.Id.NAME,
+    include = JsonTypeInfo.As.EXISTING_PROPERTY,
+    property = "{{propertyName}}",
+    visible = true
+)
+@JsonSubTypes({
+{{#mappedModels}}
+    @JsonSubTypes.Type(value = {{modelName}}.class, name = "{{mappingName}}"){{^-last}},{{/-last}}
+{{/mappedModels}}
+})
```
Comment (gh-yzou) [inline server-templates/typeInfoAnnotation.mustache:29]: new line

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+    @Test
+    public void testLongCatalogName() throws JsonProcessingException {
+        String longName = "a".repeat(1000);
+        Catalog catalog = new Catalog(
+                Catalog.TypeEnum.INTERNAL,
+                longName,
+                new CatalogProperties(TEST_LOCATION),
+                null);
+        String json = mapper.writeValueAsString(catalog);
```
Comment (gh-yzou) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:191]: it seems all tests below are doing round trip testing, i think we can probably extract a testRoundTrip method that takes a Catalog the ensures all necessary fields  for the catalog is serialized and deserialized properly, includes name, type, properties etc.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
+# Catalog Serialization Tests
+
+This directory contains test cases for validating the serialization and deserialization of Catalog objects in the Polaris management model.
+
+## Test Coverage
+
```
Comment (gh-yzou) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/README.md:6]: This readme seems very specific to the catalogSerialization and deserialization test, not a general readme for all admin/model tests. I think the test suite name its-self is pretty self-explaining about the coverage, and you can probably move the details about the test responsibility, coverage, edge case, error handling to the particular test suite as java doc

**Comment by gh-yzou** [comment]
@vedansh-5 would you mind put a screen shot about how it looks like before ?

**Comment by vedansh-5** [comment]
> @vedansh-5 would you mind put a screen shot about how it looks like before ?

image

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java (hunk 4 above)
Comment (dimas-b) [reply api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:141]: Polaris builds on 21+ IIRC

Comment (dimas-b) [reply api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:141]: With `""""` string literals you can avoid escaping `"` in JSON, I hope.

File: .vscode/settings.json (hunk 7 above)
Comment (dimas-b) [reply .vscode/settings.json:4]: Please remove this file from the PR

File: .gitignore [hunk 11]
```diff
@@ -92,3 +92,6 @@ hs_err_pid*
 
 # macOS
 *.DS_Store
+
+# vscode
+.vscode/settings.json
```
Comment (gh-yzou) [inline .gitignore:97]: new line

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+  public void testInvalidJsonDeserialization() {
+    String invalidJson = "{ invalid json }";
+    assertThrows(JsonProcessingException.class, () -> mapper.readValue(invalidJson, Catalog.class));
+  }
+
+  // Tests handling of empty string values in catalog fields
+  @Test
+  public void testCatalogWithEmptyFields() throws JsonProcessingException {
+    String json =
+        """
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:173]: @snazy I see that the `polaris-client` plugin applied to this source, requires java 11... Is that a hard requirement? Polaris server builds for java 21 apparently.

File: server-templates/typeInfoAnnotation.mustache [hunk 13]
```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (dimas-b) [inline server-templates/typeInfoAnnotation.mustache:2]: Please use the same header style as in other `.mustache` files (e.g. `server-templates/queryParams.mustache`)

File: build.gradle.kts [hunk 14]
```diff
@@ -126,6 +126,20 @@ tasks.named<RatTask>("rat").configure {
 
   excludes.add("**/kotlin-compiler*")
   excludes.add("**/build-logic/.kotlin/**")
+
+  excludes.add("docs/**") // Added this line to bypass adding header in docs file
+
+  excludes.addAll(
```
Comment (gh-yzou) [inline build.gradle.kts:132]: can you add a comment here about why we are excluding those files here ?

**Review by gh-yzou — APPROVED** [review]
I don't have further question, left some minor comments, please make sure @dimas-b 's concern is resolved.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+   * @return The deserialized catalog for additional assertions if needed
+   */
+  private Catalog verifyRoundTrip(Catalog original) throws JsonProcessingException {
+    // Perform serialization and deserialization
+    String json = mapper.writeValueAsString(original);
+    Catalog deserialized = mapper.readValue(json, Catalog.class);
+
+    // Verify required fields
+    assertEquals(original.getType(), deserialized.getType(), "Catalog type should match");
+    assertEquals(original.getName(), deserialized.getName(), "Catalog name should match");
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:81]: `Catalog` has an `equals()` method, let's do `assertThat(deserialized).isEqualTo(original)`

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+            new CatalogProperties(TEST_LOCATION),
+            storageConfig);
+    verifyRoundTrip(catalog);
+  }
+
+  // Update testCatalogWithNullFields - already using proper verification
+  @Test
+  public void testCatalogWithNullFields() throws JsonProcessingException {
+    Catalog catalog = new Catalog(Catalog.TypeEnum.INTERNAL, null, null, null);
+    verifyRoundTrip(catalog);
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:159]: I think having all those individual test methods that call into `verifyRoundTrip()` is confusing. Let's have one `@ParmeterizedTest` method with a `@MethodSource` to general all those input permutations.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+  }
+
+  // Tests error handling for invalid JSON input during deserialization
+  @Test
+  public void testInvalidJsonDeserialization() {
+    String invalidJson = "{ invalid json }";
+    assertThrows(JsonProcessingException.class, () -> mapper.readValue(invalidJson, Catalog.class));
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:166]: This is testing Jackson, not `Catalog`. I suggest removing this test.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+  }
+
+  // Tests handling of empty string values in catalog fields
+  @Test
+  public void testCatalogWithEmptyFields() throws JsonProcessingException {
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:171]: Actually, what value does this test add on top of `verifyRoundTrip()`? We can have empty strings as one of the input options to `verifyRoundTrip()` in the parameterized test discussed above.

Side benefit: we do not need to worry about `"""` String literals in that case :)

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+
+    Catalog catalog = mapper.readValue(json, Catalog.class);
+    assertEquals("", catalog.getName());
+    assertEquals("", catalog.getProperties().getDefaultBaseLocation());
+  }
+
+  // Update testSpecialCharacters
+  @Test
+  public void testSpecialCharacters() throws JsonProcessingException {
+    String specialName = "test\"catalog";
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:195]: Let's make it one of the input options to the parameterized test for `verifyRoundTrip()`

File: build.gradle.kts [hunk 20]
```diff
@@ -126,6 +126,20 @@ tasks.named<RatTask>("rat").configure {
 
   excludes.add("**/kotlin-compiler*")
   excludes.add("**/build-logic/.kotlin/**")
+
+  excludes.add("docs/**") // Added this line to bypass adding header in docs file
```
Comment (dimas-b) [inline build.gradle.kts:130]: Is this change necessary?

**Review by dimas-b — COMMENTED** [review]
The fix approach LGTM :+1: 

I'd suggest refactoring the tests, though. Two main points:
1. Use AssertJ
2. Use parameterized test for different input value peremutations.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+        // Long name
+        new TestCase("Long name")
+            .withCatalog(new Catalog(
+                Catalog.TypeEnum.INTERNAL,
+                "a".repeat(1000),
+                new CatalogProperties(TEST_LOCATION),
+                null)),
+
+        // Unicode characters
+        new TestCase("Unicode characters")
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:126]: nit: the description string is sufficient, why comment?

File: api/management-model/build.gradle.kts [hunk 22]
```diff
@@ -29,6 +29,9 @@ dependencies {
   compileOnly(libs.jakarta.annotation.api)
   compileOnly(libs.jakarta.validation.api)
   compileOnly(libs.swagger.annotations)
+  testImplementation(platform("org.junit:junit-bom:5.9.2"))
```
Comment (dimas-b) [inline api/management-model/build.gradle.kts:32]: Please use `testImplementation(platform(libs.junit.bom))` - the version is defined in `gradle/libs.versions.toml`

Comment (dimas-b) [reply api/management-model/build.gradle.kts:32]: nit: maybe add a new line above?

File: api/management-model/build.gradle.kts [hunk 23]
```diff
@@ -54,12 +57,26 @@ openApiGenerate {
   serverVariables = mapOf("basePath" to "api/v1")
 }
 
-listOf("sourcesJar", "compileJava").forEach { task ->
-  tasks.named(task) { dependsOn("openApiGenerate") }
```
Comment (dimas-b) [inline api/management-model/build.gradle.kts:58]: why move this piece of code?

File: api/management-model/build.gradle.kts [hunk 24]
```diff
@@ -54,12 +58,26 @@ openApiGenerate {
...
+tasks.named<Test>("test") { dependsOn("openApiGenerate") }
+
+tasks.withType<Test> {
+  useJUnitPlatform()
+  testLogging {
+    events("passed", "skipped", "failed")
+    showStandardStreams = true
+  }
+  dependsOn("openApiGenerate")
+}
```
Comment (dimas-b) [inline api/management-model/build.gradle.kts:83]: This change is not necessary AFAIK

File: api/management-model/build.gradle.kts (hunk 23 above)
Comment (dimas-b) [reply api/management-model/build.gradle.kts:58]: I believe only the new `testImplementation` depedencies and `test { java { srcDir("src/test/java") } }` are needed. All other build script changes seem to be superfluous - please remove them to keep the scripts lean.

File: spec/polaris-management-service.yml [hunk 25]
```diff
@@ -796,6 +796,13 @@ components:
             - INTERNAL
             - EXTERNAL
           description: the type of catalog - internal or external
+          x-java-field-annotations: >-
```
[the remark this answers is not in the document]
Comment (dimas-b) [reply spec/polaris-management-service.yml:799]: Please remove this change. I do not think it is necessary.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+ * </ul>
+ *
+ * Error handling coverage:
+ *
+ * <ul>
+ *   <li>Invalid JSON input
+ *   <li>Malformed JSON structure
+ *   <li>Invalid enum values
+ *   <li>Edge cases like very long catalog names
+ * </ul>
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:54]: nit: I'd also remove this javadoc. Test methods should self-describing and keeping test javadoc in sync with test code is overhead.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+
+  private ObjectMapper mapper;
+  private static final String TEST_LOCATION = "s3://test/";
+  private static final String TEST_CATALOG_NAME = "test-catalog";
+  private static final String TEST_ROLE_ARN = "arn:aws:iam::123456789012:role/test-role";
+
+  @BeforeEach
+  public void setUp() {
+    mapper = new ObjectMapper();
+    mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:66]: Why do we need this feature?

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 28]
```diff
@@ -0,0 +1,197 @@
...
+   * @param original The catalog object to test
+   * @return The deserialized catalog for additional assertions if needed
+   */
+  private Catalog verifyRoundTrip(Catalog original) throws JsonProcessingException {
+    // Perform serialization and deserialization
+    String json = mapper.writeValueAsString(original);
+    Catalog deserialized = mapper.readValue(json, Catalog.class);
+
+    // Compare the content instead of direct object equality
+    assertThat(deserialized.getType()).isEqualTo(original.getType());
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:82]: why not simply `assertThat(deserialized).isEqualTo(original)`?

File: api/management-model/build.gradle.kts [hunk 29]
```diff
@@ -54,12 +58,26 @@ openApiGenerate {
   serverVariables = mapOf("basePath" to "api/v1")
 }
 
-listOf("sourcesJar", "compileJava").forEach { task ->
-  tasks.named(task) { dependsOn("openApiGenerate") }
-}
-
 sourceSets {
   main { java { srcDir(project.layout.buildDirectory.dir("generated/src/main/java")) } }
+  test { java { srcDir("src/test/java") } }
```
Comment (dimas-b) [inline api/management-model/build.gradle.kts:63]: Actually, I think even this change is not necessary - it's default in gradle. Only the new `testImplementation` deps are needed.

File: build.gradle.kts (hunk 14 above)
Comment (dimas-b) [reply build.gradle.kts:132]: I believe this change is not necessary, let's roll it back.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java (hunk 12 above)
Comment (snazy) [reply api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java]: I personally would not mind to bump everything to Java 21

File: .gitignore [hunk 30]
```diff
@@ -92,3 +92,6 @@ hs_err_pid*
 
 # macOS
 *.DS_Store
+
+# vscode
```
Comment (snazy) [inline .gitignore:96]: This change's unrelated to the scope of the PR - please remove.

File: api/management-model/build.gradle.kts [hunk 31]
```diff
@@ -29,6 +29,10 @@ dependencies {
   compileOnly(libs.jakarta.annotation.api)
   compileOnly(libs.jakarta.validation.api)
   compileOnly(libs.swagger.annotations)
+
+  testImplementation(platform(libs.junit.bom))
+  testImplementation("org.junit.jupiter:junit-jupiter")
+  testImplementation("com.fasterxml.jackson.core:jackson-databind:2.15.2")
```
Comment (snazy) [inline api/management-model/build.gradle.kts:35]: Use the Jackson-bom and avoid version declarations here.

File: api/management-model/build.gradle.kts (hunk 23 above)
Comment (snazy) [reply api/management-model/build.gradle.kts:58]: Agree, this change's unrelated to the scope of the PR - please revert.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 32]
```diff
@@ -0,0 +1,172 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.admin.model;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (snazy) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:22]: Avoid wildcard imports + prefer assertJ

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 33]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    @Override
+    public String toString() {
+      return description;
+    }
+  }
+
+  @Test
+  public void testMalformedJson() {
```
Comment (snazy) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:167]: How's this test case related to the scope of the PR?

File: build.gradle.kts (hunk 14 above)
Comment (snazy) [reply build.gradle.kts:130]: This change's unrelated to the scope of the PR - please remove.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 34]
```diff
@@ -0,0 +1,172 @@
...
+                                Catalog.TypeEnum.INTERNAL,
+                                TEST_CATALOG_NAME,
+                                new CatalogProperties(TEST_LOCATION),
+                                new AwsStorageConfigInfo(
+                                    arn, StorageConfigInfo.StorageTypeEnum.S3))));
+
+    return Stream.concat(basicCases, arnCases);
+  }
+
+  private static class TestCase {
```
Comment (snazy) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:147]: Why this class and not just parameterized-test arguments?

**Review by snazy — CHANGES_REQUESTED** [review]
The approach in general is fine, however I don't see how the test cases ensure that the type-field itself is only serialized once.

Beside that there are a bunch of changes that are out of scope of the PR - those need to be removed.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java (hunk 32 above)
Comment (dimas-b) [reply api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:22]: good catch :+1:

File: .gitignore [hunk 35]
```diff
@@ -91,4 +91,4 @@ logs/
 hs_err_pid*
 
 # macOS
-*.DS_Store
+*.DS_Store
```
Comment (dimas-b) [inline .gitignore:94]: Could you please revert this change? It does not look necessary.

File: build.gradle.kts (hunk 14 above)
Comment (dimas-b) [reply build.gradle.kts:130]: ping: please remove this change.

**Comment by vedansh-5** [comment]
Hi @dimas-b ,

I noticed that I was removed as an assignee from #955 after I didn’t respond to the reviews on time. I sincerely apologize for the delay, I had exams going on, and it was a bit difficult to manage everything alongside my studies.

I’ve now gone through the reviews, made the necessary changes, and updated the PR accordingly. Please let me know if anything else needs to be addressed. I’d love to continue contributing and really appreciate your time and patience!

Thanks

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 36]
```diff
@@ -0,0 +1,132 @@
...
+  public void setUp() {
+    mapper = new ObjectMapper();
+    // Mapper to use the concrete Catalog class
+    mapper.addMixIn(PolarisCatalog.class, Catalog.class);
+  }
+
+  private Catalog verifyRoundTrip(Catalog original) throws JsonProcessingException {
+    String json = mapper.writeValueAsString(original);
+    int typeFieldCount = json.split("\"type\"").length - 1;
+    assertThat(typeFieldCount)
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:47]: This is too complicated, IMHO. The JSON string form of `Catalog` is stable. It will not change from one execution to another. I'd prefer to assert the exact string JSON representation in a test method dedicated to this case and leave this assertion out of the "round trip" tests.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 37]
```diff
@@ -0,0 +1,132 @@
...
+        .as("type field should appear exactly once for %s catalog", original.getType())
+        .isEqualTo(1);
+
+    Catalog deserialized = mapper.readValue(json, Catalog.class);
+
+    // Compare entire object recursively in one assertion
+    assertThat(deserialized).isEqualToComparingFieldByFieldRecursively(original);
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:54]: why not simply `assertThat(deserialized).isEqualTo(original)`?

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java (hunk 34 above)
Comment (dimas-b) [reply api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:110]: I agree with @snazy .

Please consider producing individual `Catalog` objects as test method arguments (and remove `TestCase`).

If you prefer to have hand-written test case descriptions, please consider using `ArgumentSet` (from JUnit5).

**Review by dimas-b — COMMENTED** [review]
Some more comments... Also CI appears to have failed :shrug:

File: .gitignore (hunk 35 above)
Comment (dimas-b) [reply .gitignore:94]: ping: please remove this change

File: .vscode/settings.json (hunk 7 above)
Comment (dimas-b) [reply .vscode/settings.json:4]: ping: Please remove this file from the PR

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 38]
```diff
@@ -0,0 +1,125 @@
...
+   * Helper method to verify round-trip serialization/deserialization of Catalog objects. Ensures
+   * all fields are preserved correctly through the process.
+   *
+   * @param original The catalog object to test
+   * @return The deserialized catalog for additional assertions if needed
+   */
+  private Catalog verifyRoundTrip(Catalog original) throws JsonProcessingException {
+    String json = mapper.writeValueAsString(original);
+    Catalog deserialized = mapper.readValue(json, Catalog.class);
+    assertThat(deserialized).usingRecursiveComparison().isEqualTo(original);
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:53]: Why not simply assertThat(deserialized).isEqualTo(original)?

Do you think the `.equals()` method in `Catalog` does not take nested objects into account?

**Review by dimas-b — APPROVED** [review]
Thanks for you patience @vedansh-5 !

The current state of this PR LGTM :+1: 

@flyrain @eric-maynard @dennishuo : Please double check mustache changes.

**Comment by dimas-b** [comment]
@snazy : Could you have another look too?

**Review by dimas-b — APPROVED** [review]

**Comment by vedansh-5** [comment]
@snazy : I have worked on the proposed changes, could you please take another look at them. Thankyou for your feedback.

**Comment by vedansh-5** [comment]
Hi @snazy,
I have worked on the requested changes based on your feedback. When you have some time, could you please take another look? Your review would be greatly appreciated.
Thanks!

**Comment by vedansh-5** [comment]
> Are there test cases ensuring that the type-field itself is only serialized once?

I have added test case for the same.

**Comment by snazy** [comment]
> > Are there test cases ensuring that the type-field itself is only serialized once?
> 
> I have added test case for the same.

And where? I don't see any executed test case that checks it.

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 39]
```diff
@@ -0,0 +1,144 @@
...
+    return deserialized;
+  }
+
+  @ParameterizedTest(name = "{0}")
+  @MethodSource("catalogTestCases")
+  public void testCatalogSerialization(String description, Catalog catalog)
+      throws JsonProcessingException {
+    verifyRoundTrip(catalog);
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:61]: `verifyRoundTrip` is called from only this one place. Please inline it.

**Review by dimas-b — COMMENTED** [review]
Good point by @snazy 

@vedansh-5 : IIRC, we have a test validating the serialized JSON string at some point... I guess it was lost in the refactrorings. Cf. https://github.com/apache/polaris/pull/955#discussion_r1945275773

Please re-add a test method that asserts the exact JSON string form of a java `Catalog` object.

**Comment by vedansh-5** [comment]
> Good point by @snazy
> 
> @vedansh-5 : IIRC, we have a test validating the serialized JSON string at some point... I guess it was lost in the refactrorings. Cf. [#955 (comment)](https://github.com/apache/polaris/pull/955#discussion_r1945275773)
> 
> Please re-add a test method that asserts the exact JSON string form of a java `Catalog` object.

I have added the test method and using ```verifyRoundTrip``` as inline

File: api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java [hunk 40]
```diff
@@ -0,0 +1,162 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class CatalogSerializationTest {
+
+  @Test
+  public void testJsonFormat() throws JsonProcessingException {
```
Comment (dimas-b) [inline api/management-model/src/test/java/org/apache/polaris/core/admin/model/CatalogSerializationTest.java:35]: please move it down - together with the other test method

**Review by dimas-b — APPROVED** [review]

**Comment by vedansh-5** [comment]
@snazy: I made the requested changes. When you have some time, could you please take another look?
Thanks!

**Comment by vedansh-5** [comment]
@snazy : Follow-up for the same message

> @snazy: I made the requested changes. When you have some time, could you please take another look? Thanks!

**Review by eric-maynard — APPROVED** [review]
